File: read.py
#!/usr/bin/env python3

import sys
import zlib
from urllib.parse import unquote
import xml.etree.ElementTree as ET
import base64

PNG_HEAD = b"\x89\x50\x4e\x47\x0d\x0a\x1a\x0a" # Magic PNG header bytes
PNG_IEND = b"\x49\x45\x4E\x44"
PNG_ZTXT = b"\x7A\x54\x58\x74" # info is stored in zTXT section


"""
    It might be good to peruse the PNG spec:
    https://www.w3.org/TR/PNG-Structure.html
"""

"""
    This file takes a single argument: path to a PNG file you expect
    contains an embedded draw.io diagram data. This only happens if 'Include a copy of my diagram' was checked when exporting

    This will Read the PNG file, and print the resulting diagram data to stdout.
"""

def inflate(b,b64=False):
    """~2016 draw.io started compressing 'using standard deflate'
        https://about.draw.io/extracting-the-xml-from-mxfiles/
        experience has shown this is deflate WITH NO HEADER
    """
    if b64: # optional, additionally base64 decode
        b = base64.b64decode(b)
    return unquote(zlib.decompress(b,-15).decode('utf8'))

def valid_png(f):
    head = f.read(8)
    if head != PNG_HEAD:
        print("invalid PNG")
        sys.exit(1)

def read_section(f):
    """Sections/chunks are
        4-byte LENGTH
        4-byte SECTION
        <LENGTH> CONTENTS
        4-byte CRC
    """
    length = int.from_bytes(f.read(4), byteorder='big',signed=False) # in bytes
    sectype = f.read(4)

    contents = f.read(length)
    f.read(4) # skip CRC int
    # or f.seek(4, 1) to just move 4 bytes forward from current

    return sectype,contents


def main():
    ztxt = {}
    with open(sys.argv[1], "rb") as f:
        valid_png(f)
        while f.read(1) != b'':
            f.seek(-1,1)
            sectype,contents = read_section(f)
            if sectype == PNG_IEND:
                break
            elif sectype == PNG_ZTXT:
                idx=0
                while contents[idx] != 0:
                    idx += 1
                keyname = contents[:idx].decode('ascii')
                data = contents[idx+2:] # skip two NUL bytes
                ztxt[keyname] = data

    """
        Basically, the PNG should have a zTXT section, which is
        itself like a key/value store. It should have a key of
        mxGraphModel. The data for that key is compressed XML,
        which has node with text in it. That text is base64 encoded,
        and again compressed. So undoing all of that we get XML again,
        of the true mxGraphModel. That graph model is what draw.io
        uses to recompute the diagram. You can even paste it in 
        Extras > Edit Diagram.
    """

    xml = inflate(ztxt['mxGraphModel'])
    mxfile = ET.fromstring(xml)[0].text
    diagram = inflate(mxfile,b64=True)
    print(diagram)

if __name__ == "__main__":
    main()

File: read.go
package main

import (
	"fmt"
	"os"

	// byte array operations
	"bytes"
	"io"
	"io/ioutil"

	//byte parsing
	"encoding/binary"

	// drawio-related parsing
	"compress/flate"
	"encoding/base64"
	"encoding/xml"
	"net/url"
)

const (
	pngHead = "\x89PNG\r\n\x1a\n"
	ztxtHead = "zTXt"
	endHead = "IEND"
)

type Chunk struct {
	Length uint32
	Type   string
	Data   []byte
	Crc32  []byte
}

type ZtChunk struct {
	Name string
	Text string
}

type MXFile struct {
	Diagram string `xml:"diagram"`
}


func (c *Chunk) readSection(f io.Reader) {
	buf4 := make([]byte, 4)

	io.ReadFull(f, buf4)
	c.Length = binary.BigEndian.Uint32(buf4)

	io.ReadFull(f, buf4)
	c.Type = string(buf4)

	databuf := make([]byte, c.Length)
	io.ReadFull(f, databuf)
	c.Data = databuf

	io.ReadFull(f, buf4)
	c.Crc32 = buf4
}

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func deflate(data []byte) (string, error) {
	reader := flate.NewReader(bytes.NewReader(data))
	defer reader.Close()

	uncompressed, err := ioutil.ReadAll(reader)
	return string(uncompressed), err
}

func parseZTXT(c *Chunk) ZtChunk {
	var z ZtChunk
	nul_pos := bytes.Index(c.Data, []byte{0})
	z.Name = string(c.Data[:nul_pos])
	// there are two NUL bytes. One for separation, other to mark compression
	data, err := deflate(c.Data[nul_pos+2:])
	check(err)
	z.Text = data

	return z
}

func validPNG(f io.Reader) (bool, error) {
	head := make([]byte, 8)
	_, err := io.ReadFull(f, head)
	if err != nil {
		return false, err
	}
	return string(head) == pngHead, nil
}



func main() {
	if len(os.Args) < 2 {
		fmt.Println("Argument required: path to PNG file")
		os.Exit(1)
	}
	filename := os.Args[1]
	file, err := os.Open(filename)
	check(err)
	defer file.Close()

	valid, err := validPNG(file)
	check(err)
	if !valid {
		fmt.Println("Not valid PNG file")
		os.Exit(1)
	}

	for chunk := new(Chunk); chunk.Type != endHead; chunk.readSection(file) {
		if chunk.Type == ztxtHead {
			z := parseZTXT(chunk)
			if z.Name == "mxGraphModel" {
				data := z.Text
				data, _ = url.QueryUnescape(data)
				var mxfile MXFile
				xml.Unmarshal([]byte(data), &mxfile)
				decoded, err := base64.StdEncoding.DecodeString(mxfile.Diagram)
				check(err)

				uncompressed, err := deflate(decoded)
				check(err)
				final, _ := url.QueryUnescape(uncompressed)
				fmt.Println(final)
				os.Exit(0)
			}
			break
		}
	}
}
